File: data-structures/ruby/stack/stack_spec.rb
require_relative 'stack'
require_relative 'stack_node'

RSpec.describe Stack do
  context 'initial empty stack' do
    stack = Stack.new
    it 'is empty upon initialization' do
      expect( stack.isEmpty? ).to be( true )
    end

    it 'raises error if try to pop upon initialization' do
      expect{ stack.pop }.to raise_error( StandardError, 'Stack is empty, cannot pop!' )
    end

    it 'returns nil when peeking' do
      expect( stack.peek ).to eq( nil )
    end
  end

  context 'adding nodes to stack' do
    stack = Stack.new
    node1 = StackNode.new( data: 1 )

    it 'is not empty after pushing a node' do
      stack.push( node1 )
      expect( stack.isEmpty? ).to be( false )
    end

    it 'returns entire top node when peeking' do
      stack.push( node1 )
      expect( stack.peek ).to eq( node1 )
    end
  end

  context 'removing nodes from stack' do
    stack = Stack.new
    node1 = StackNode.new( data: 1 )
    node2 = StackNode.new( data: 2 )
    stack.push( node1 )
    stack.push( node2 )

    it 'returns the top node when popped' do
      expect( stack.pop ).to eq( node2 )
    end

    it 'returns the next node after popping when peeking' do
      expect( stack.peek ).to eq( node1 )
    end

    it 'is empty when all nodes are popped' do
      stack.pop
      expect( stack.isEmpty? ).to be( true )
    end
  end
end

File: data-structures/javascript/stack/stack_spec.js
const stack = require( './stack' );
const stackNode = require( './stack_node' );

describe ( 'Stack', () => {

  describe ( 'initial empty Stack', () => {
    const testStack = Object.create( stack );

    it ( 'is empty upon initialization', () => {
      expect( testStack.isEmpty() ).toBe( true );
    });

    it ( 'returns error message if try to pop upon initialization', () => {
      expect( testStack.pop() ).toEqual( 'Stack is empty, cannot pop!' )
    });

    it ( 'returns null when peeking', () => {
      expect( testStack.peek() ).toEqual( null );
    })
  });

  describe ( 'adding nodes to stack', () => {
    const testStack = Object.create( stack );
    const node1 = Object.create( stackNode );
    node1.initialize( 1 );

    it ( 'is not empty after pushing a node', () => {
      testStack.push( node1 );
      expect( testStack.isEmpty() ).toBe( false );
    });

    it ( 'returns top node when peeking', () => {
      testStack.push( node1 );
      expect( testStack.peek() == node1 ).toBeTruthy();
    });
  });

  describe ( 'removing nodes from stack', () => {
    const testStack = Object.create( stack );
    const node1 = Object.create( stackNode );
    const node2 = Object.create( stackNode );

    node1.initialize( 1 );
    node2.initialize( 2 );
    testStack.push( node1 );
    testStack.push( node2 );

    const node2Next = node2.next;

    it ( 'returns the top node when popped', () => {
      expect( testStack.pop() == node2 ).toBeTruthy();
    });

    it ( 'returns the next node after popping when peekings', () => {
      expect( testStack.peek()  == node1 && node1 == node2.next ).toBeTruthy();
    });

    it ( 'is empty when all nodes are popped', () => {
      testStack.pop();
      expect( testStack.isEmpty() ).toBe( true );
    });
  });

});
